### ISSUE TITLE
Exceptions are stringified too early

### ISSUE THREAD

**Issue by nilskp** [outr/scribe#44]
`LogRecord` has no trace of exceptions, except in the message.

This makes it hard to implement handlers that are interested in exceptions only.

**Comment by darkfrog26** [comment]
A very good point. Let me think a little more about this. Have any suggested solution?

**Comment by nilskp** [comment]
If I'm not mistaken, most APIs simply have an optional exception has part of the log record.

**Comment by darkfrog26** [comment]
I'm aware, but that's not incredibly flexible. I'm currently prototyping something a bit more powerful that would allow any type to be used.

**Comment by darkfrog26** [comment]
@nilskp Take a look at https://github.com/outr/scribe/blob/master/core/shared/src/main/scala/scribe/LogRecord.scala#L8

You can access the original instance being logged (in the case of a `Throwable` you'll have a reference to the object directly).  Is that sufficient?  If this works I can add some additional functionality to easily filter and map this way.

**Comment by nilskp** [comment]
@darkfrog26 Generally, it feels like a code smell to use `Any`. It seems like you're trying to cover a not yet existing use case, which could end up being a YAGNI.

Exceptions are first-class citizens when logging. It seems like they should be expressed explicitly.

More specifically, I don't think it's possible to use e.g. `.warn("Blimey!", exception)` here, at least not when using `SimpleLogRecord`, or am I misreading?

**Comment by darkfrog26** [comment]
In Scribe that works, but emits two LogRecords.

**Comment by nilskp** [comment]
With identical content, except for String and Exception?

**Comment by darkfrog26** [comment]
I don't necessarily disagree about the use of `Any`, but I think this is one of those cases where it does make sense.  I have projects where I could see this being useful.  For example, if doing database logging I might deal with specific instances of domain-specific objects distinctly (for example if I log a `User` instance).

**Comment by darkfrog26** [comment]
See https://github.com/outr/scribe/blob/master/macros/shared/src/main/scala/scribe/Macros.scala#L78

**Comment by nilskp** [comment]
Wouldn't that lead to two log statements, rather than one?

**Comment by darkfrog26** [comment]
Yes, but I generally find that a better representation anyway.

**Comment by darkfrog26** [comment]
I'm very open to considering alternatives, but thus far I've had no negative reaction to doing it that way.  Please let me know if you have a problem with this as I'd like to better understand.

**Comment by nilskp** [comment]
Specifically my use case right now is sending exceptions to Sentry. Having the message separated from the exception can probably not be reconciled without issues.

Not all logging goes into a linear text file. If it's logged together, it should be encapsulated into a single conceptual record.

That's my opinion anyways. :-)

**Comment by darkfrog26** [comment]
If you really care about keeping it together, why not just wrap the exception with another exception including the message?

**Comment by darkfrog26** [comment]
So, instead of:

```scala
scribe.error("Oh no!", exc)
```

It would be something like:

```scala
scribe.error(new RuntimeException("Oh no!", exc)
```

**Comment by darkfrog26** [comment]
I understand and can appreciate your perspective, but in my use-case I'm almost always just logging the exception directly:

```scala
scribe.error(exc)
```

I let the exception explain itself.  This is why I'm trying to better understand since this doesn't fall into the norm of how I typically interact with logging.

**Comment by nilskp** [comment]
But why? A log record ought to be a single record of what I've logged.

Anyways, I'm spending too much time on this already.

Scribe is not essential to what I'm doing. I would have incorporated it if it was easy, but now it's looking like too much work.

Hope you come up with a good solution.

👍

**Comment by darkfrog26** [comment]
I'd like to work with you as I believe that Scribe should be faster and more flexible than any other Scala logging libraries, so if you're willing to put in a little more effort I'm sure we can come to an ideal solution.

**Comment by nilskp** [comment]
I would just say that if you want to have this adopted by more people, it probably needs to follow certain conventions and would need to fill existing use cases, such as external adapters etc.
It's a non-starter for me, and probably a lot of people, if the semantics of logging is changed, i.e. `.log(string, exception)` leads to two log records.
If I were you, I'd look into supporting at least a couple of external ecosystem use cases, such as e.g. Sentry.io. Not necessarily for publishing and maintaining, but to see if you can fill those use cases easily.

**Comment by darkfrog26** [comment]
@nilskp is this the only roadblock from you using Scribe in your project?  This isn't an unreasonable request.  I should be able to accommodate.  Let me know if there are any other issues you see.

I'm going to be doing release 2.3.0 early next week and should be able to get this into it.

**Comment by darkfrog26** [comment]
@nilskp I've updated the system to, hopefully, accommodate all of your concerns:

1.) I've updated the system to only create one log record for message + throwable mimicking how log4j does it.
2.) I've removed `Any` from `LogRecord` and in-place have created a generic type to allow all the advantages of both retaining the original type as well as deferring stringification until actually logged. See https://github.com/outr/scribe/blob/master/core/shared/src/main/scala/scribe/LogRecord.scala#L9

**Comment by darkfrog26** [comment]
If this alleviates all your concerns please confirm and close this ticket.  Tentatively this will go out in 2.3.0 on Monday.

**Comment by nilskp** [comment]
I've just started looking at this, and Sentry was the first blocker.

I won't have time to look at this until sometime next week, but I don't see why it wouldn't work.

Thanks for addressing this.

**Comment by darkfrog26** [comment]
@nilskp no worries man, and I appreciate the feedback. 

My goal is to make the best all-around logger for Scala, so please keep the feedback coming. :)

**Comment by darkfrog26** [comment]
This functionality was just released with 2.3.0
